module/os/linux/zfs/zpl_xattr.c

```diff
@@ -1384,6 +1407,410 @@ static xattr_handler_t zpl_xattr_acl_default_handler = {
+	}
+
+	*xattrbuf++ = htonl(acep->a_type);
+	*xattrbuf++ = htonl(acep->a_flags & NFS41_FLAGS);
+	*xattrbuf++ = htonl(iflag);
+	*xattrbuf++ = htonl(acep->a_access_mask);
+	*xattrbuf++ = htonl(who);
+
+	return (0);
+}
+
+static int
+zfsacl_to_nfsacl41i(const vsecattr_t vsecp, u32 *xattrbuf)
+{
+	int i, error = 0;
+	ace_t *acep = NULL;
+
+	*xattrbuf++ = htonl(vsecp.vsa_aclflags);
+	*xattrbuf++ = htonl(vsecp.vsa_aclcnt);
+
+	for (i = 0; i < vsecp.vsa_aclcnt; i++, xattrbuf += ACE4ELEM) {
+		acep = vsecp.vsa_aclentp + (i * sizeof (ace_t));
+
+		error = acep_to_nfsace4i(acep, xattrbuf);
+		if (error)
+			break;
+	}
+
+	return (error);
+}
+
+static int
+nfsace4i_to_acep(const u32 *xattrbuf, ace_t *acep)
+{
+	u32 iflag, id;
+
+	acep->a_type = ntohl(*(xattrbuf++));
+	acep->a_flags = ntohl(*(xattrbuf++)) & NFS41_FLAGS;
+	iflag = ntohl(*(xattrbuf++));
+	acep->a_access_mask = ntohl(*(xattrbuf++));
+	id = ntohl(*(xattrbuf++));
+
+	if (iflag & ACEI4_SPECIAL_WHO) {
+		switch (id) {
+		case ACE4_SPECIAL_OWNER:
+			acep->a_flags |= ACE_OWNER;
+			acep->a_who = -1;
+			break;
+
+		case ACE4_SPECIAL_GROUP:
+			acep->a_flags |= (ACE_GROUP | ACE_IDENTIFIER_GROUP);
+			acep->a_who = -1;
+			break;
+
+		case ACE4_SPECIAL_EVERYONE:
+			acep->a_flags |= ACE_EVERYONE;
+			acep->a_who = -1;
+			break;
+
+		default:
+			dprintf("Unknown id 0x%08x\n", id);
+			return (-EINVAL);
+		}
+	} else {
+		acep->a_who = id;
+	}
+
+	return (0);
+}
+
+static int
+nfsacl41i_to_zfsacl(const u32 *xattrbuf, size_t bufsz, vsecattr_t *vsecp)
+{
+	int error;
+	int i;
+
+	vsecp->vsa_aclflags = ntohl(*(xattrbuf++));
+	vsecp->vsa_aclcnt = ntohl(*(xattrbuf++));
+	bufsz -= (2 * sizeof (u32));
+	vsecp->vsa_aclentsz = vsecp->vsa_aclcnt * sizeof (ace_t);
+
+	if (bufsz != (vsecp->vsa_aclcnt * ACE4SIZE)) {
+		dprintf("Embedded ACL count [%d] is larger than "
+		    "can fit in provided buffer size: %zu\n",
+		    vsecp->vsa_aclcnt, bufsz);
+		return (-ERANGE);
+	}
+
+	vsecp->vsa_aclentp = kmem_alloc(vsecp->vsa_aclentsz, KM_SLEEP);
+
+	for (i = 0; i < vsecp->vsa_aclcnt; i++, xattrbuf += ACE4ELEM) {
+		ace_t *acep = vsecp->vsa_aclentp + (i * sizeof (ace_t));
+
+		error = nfsace4i_to_acep(xattrbuf, acep);
+		if (error) {
+			kmem_free(vsecp->vsa_aclentp, vsecp->vsa_aclentsz);
+			return (error);
+		}
+	}
+
+	return (0);
+}
+
+static int
+__zpl_xattr_nfs41acl_get(struct inode *ip, const char *name,
+    void *buffer, size_t size)
+{
+	vsecattr_t vsecp = {0};
+	cred_t *cr = CRED();
+	int ret, fl;
+	size_t xdr_size;
+
+	if (ITOZSB(ip)->z_acl_type != ZFS_ACLTYPE_NFSV4)
+		return (-EOPNOTSUPP);
+
+	if (size == 0) {
+		/*
+		 * API user may send 0 size so that we
+		 * return size of buffer needed for ACL.
+		 */
+		crhold(cr);
+		vsecp.vsa_mask = VSA_ACECNT;
+		ret = -zfs_getsecattr(ITOZ(ip), &vsecp, ATTR_NOACLCHECK, cr);
+		if (ret) {
+			return (ret);
+		}
+		crfree(cr);
+		ret = ACES_TO_XDRSIZE(vsecp.vsa_aclcnt);
+		return (ret);
+	}
+
+	if (size < ACES_TO_XDRSIZE(1)) {
+		return (-EINVAL);
```

Comment:
I wonder if `size` of `ACES_TO_XDRSIZE(0)` is really invalid here?  We might return it above if there are no ACLs, and we would return `-ENODATA` vs `-ERANGE` below, depending whether there are no ACLs vs there are some.  It might be not very practical, but would feel consistent.

Reply:
Thanks, I have removed returning `EINVAL` in case of `ACES_TO_XDRSIZE(0)`, so we would allow to run in that case and return `ENODATA` below in case ACL count is 0.

Reply:
I generally meant `if (size < ACES_TO_XDRSIZE(0)) {`, but your variant with removal should probably be fine too.